Add token ID metadata guide (#894)
lukso-network/docs: docs/learn/dapp-developer/read-asset-data.md

<!-- prettier-ignore-end -->

At this point, you should be able to identify if the digital asset is a:

- LSP7 - Token
- LSP7 - NFT
- LSP8 - NFT
- LSP8 - Collection

## Fetch the Asset Metadata

> Based on the [token type](#detect-the-token-type), the information of the [LSP4 Digital Metadata](../../standards/tokens/LSP4-Digital-Asset-Metadata.md#types-of-digital-assets) can be interpreted and attached in different ways:

Reviewer: ```suggestion
Based on the [token type](#detect-the-token-type), the information of the [LSP4 Digital Metadata](../../standards/tokens/LSP4-Digital-Asset-Metadata.md#types-of-digital-assets) can be interpreted differently:
```

- As [**global token information**](#global-token-information) of the contract (Token or LSP7 NFT)
- To each [**individual token ID**](#token-id-metadata) (LSP8 NFT or Collection)

:::tip Token ID Entries

- If your asset is a Token (`tokenType = 0`), individual metadata **cant be set**, due to the lack of token IDs.
- If your asset is an NFT (`tokenType = 1`), individual metadata **may be set in addition** to the global token information.
- If your asset is a Collection (`tokenType = 2`), the token ID-specific medatada is **mandatory**.

:::

### Global Token Information

To fetch the whole JSON file of the asset's metadata, you can simply use the [`fetchData('LSP4Metadata')`](../../tools/erc725js/classes/ERC725.md#fetchdata) method of the [`erc725js`](../../tools/erc725js/getting-started.md) library. Optionally, you can call the [`getData('LSP4Metadata')`](../../contracts/contracts/ERC725/ERC725.md#getdata) function on the asset contract and decode the information manually.

> :::info differece between get and fetch

Reviewer: ```suggestion
:::info difference between get and fetch
```

> The [`getData(...)`](../../tools/erc725js/classes/ERC725#getdata) function only retrieves the data keys values from the smart contract. In comparison, [`fetchData(...)`](../../tools/erc725js/classes/ERC725#fetchdata) will download and decode the content of `VerifiableURI` referenced within the smart contract. You can read more about `VerifiableURI` in the [LSP2 Specification](https://github.com/lukso-network/LIPs/blob/main/LSPs/LSP-2-ERC725YJSONSchema.md#verifiableuri).

Reviewer: This cam be grouped with above:

--- 

To fetch the whole JSON file of the asset's metadata, you can use one of the following 2 functions from _erc725.js_:
- [`fetchData('LSP4Metadata')`](../../tools/erc725js/classes/ERC725.md#fetchdata): this will download and decode the content of the [`VerifiableURI`](https://github.com/lukso-network/LIPs/blob/main/LSPs/LSP-2-ERC725YJSONSchema.md#verifiableuri). as JSON.
- [`getData('LSP4Metadata')`](../../contracts/contracts/ERC725/ERC725.md#getdata): this will retrieve the raw data value from the smart contract. You will then need to decode the `VerifiableURI` manually using `decodeData(...)`.

## Fetch the Asset Metadata

As described on the different asset types of the [LSP4 Digital Asset Metadata](../../standards/tokens/LSP4-Digital-Asset-Metadata.md#types-of-digital-assets) documentation, metadata can be attached:

- As **global token information** of the contract (Token or LSP7 NFT)

> - To each **individual token ID** (LSP8 NFT or Collection)

Reviewer 2: lets also make sure we include baseUri https://github.com/lukso-network/LIPs/blob/main/LSPs/LSP-8-IdentifiableDigitalAsset.md#lsp8tokenmetadatabaseuri

```js
async function fetchTokenIdMetadata() {
  // ...

  // Get the BaseURI of the storage provider
  let tokenBaseURI = await myAsset.getData(
    ERC725YDataKeys.LSP8['LSP8TokenMetadataBaseURI'],
  );
  console.log('BaseURI: ', tokenBaseURI);

  // Get the contentID or file link
  let contentID = decodedMetadata[0].value.url;
  console.log('ContentID: ', contentID);
}

fetchTokenIdMetadata();
```

:::info

The file address within the fetched and decoded `LSP4 Metadata` of the token ID might already contain the full link of the file, without using the base provider. If you are checking assets in production, please check if the content is a valid URL beforehand.

:::

:::caution

> In case the [`LSP8TokenMetadataBaseURI`](https://github.com/lukso-network/LIPs/blob/main/LSPs/LSP-8-IdentifiableDigitalAsset.md#lsp8tokenmetadatabaseuri) was not set, you will have to fall back to the gateway indicated within the `contentID`. In this example, we will use the IPFS gateway of LUKSO, as the `contentID` starts with `ipfs://`, meaning that it was uploaded using the [Interplanetary File System](https://ipfs.tech/).

Reviewer 2: Sorry, i still think we are missing something. We dont have a section showing how to get the metadata of a specific tokenId if the asset uses `LSP8TokenMetadataBaseURI`.

Instead of fetching LSP4Metadata per tokenId as in the `fetchTokenIdMetadata` function, you would need to fetch the base URI retrieved from `LSP8TokenMetadataBaseURI` to construct the metadata url as `https://mybase.uri/TOKEN_ID` which can be used for all assets

Author: Thanks for bearing with me.  🙏🏻 I just talked back to Daniel, and we should indeed split Metadata and BaseURI up into separate sections instead of combining them. 

I thought they were used for the same purpose and could be combined somehow. I will get back tomorrow and check for a good example asset. The current one does not have a BaseURI set. 🙂

Reviewer 2: perfect thanks!

```bash
npm install web3 @erc725/erc725.js @lukso/lsp-smart-contracts
```

```js
// Add the necessary imports to your JS file
import Web3 from 'web3';
import lsp8Artifact from '@lukso/lsp-smart-contracts/artifacts/LSP8IdentifiableDigitalAsset.json';
import { INTERFACE_IDS, ERC725YDataKeys } from '@lukso/lsp-smart-contracts';

const web3 = new Web3(
  new Web3.providers.HttpProvider('https://rpc.testnet.lukso.gateway.fm'),
);
// Create contract instance
const myAssetContract = new web3.eth.Contract(
  lsp8Artifact.abi,
  '<myAssetAddress>',
);
```

  </TabItem>

</Tabs>

:::caution

Make sure to adjust `<myAssetAddress>` with the actual address of the asset. You can give it a try using a sample LSP8 asset on the LUKSO Testnet: [`0x8734600968c7e7193BB9B1b005677B4edBaDcD18`](https://wallet.universalprofile.cloud/asset/0x8734600968c7e7193BB9B1b005677B4edBaDcD18?network=testnet).

:::

### Preparing the Token IDs

The [LSP8](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md) allows for different [Token ID Formats](https://github.com/lukso-network/LIPs/blob/main/LSPs/LSP-8-IdentifiableDigitalAsset.md#lsp8tokenidformat), meaning developers can specify their token IDs as:

> - `Number` of `uint256` (`TokenIdFormat`: `0` or `100`)

Reviewer: Please let's not re-write the same thing we have in some page in multiple places. Duplicate content is what will create documentation that is very hard to maintain.

Please just remove these bullet points. Developer can go to the right page to learn more.

To call the contract, you must first **prepare** your token IDs to match these standardized **Byte32 Hex Strings** based on the [`LSP8TokenIdFormat`](https://github.com/lukso-network/LIPs/blob/main/LSPs/LSP-8-IdentifiableDigitalAsset.md#lsp8tokenidformat) standardization. The **global token ID format** can be fetched from the [ERC725Y](../../standards/lsp-background/erc725.md#erc725y-generic-data-keyvalue-store) data storage using the [`getData()`](../../contracts/contracts/ERC725/ERC725.md#getdata) function:

<Tabs groupId="provider-lib">
  <TabItem value="ethers" label="ethers">

```js
// ...

// Get the global token ID format
let tokenIdFormat = parseInt(
  await myAssetContract.getData(ERC725YDataKeys.LSP8['LSP8TokenIdFormat']),
  16,
);

console.log(tokenIdFormat);
// 0
```

  </TabItem>

  <TabItem value="web3" label="web3">

```js
// ...

// Get the global token ID format
let tokenIdFormat = parseInt(
  await myAssetContract.methods
    .getData(ERC725YDataKeys.LSP8['LSP8TokenIdFormat'])
    .call(),
  16,
);

console.log(tokenIdFormat);
// 0
```

  </TabItem>

</Tabs>

Its value can be interpreted as the following:

> - `0`, `1`, `2`, `4` means the token ID format is **equal across all token IDs**

Reviewer: Again same thing, this is already mentioned in the Standard page. This page should be a How To guide, not a page to explain the standard again.

:::

<Tabs groupId="provider-lib">
  <TabItem value="ethers" label="ethers">

```js
// ...

if (isLSP8) {
  const provider = new ethers.JsonRpcProvider(
    'https://rpc.testnet.lukso.gateway.fm',
  );

  // Token ID as Bytes32 value (1)
  const tokenID =
    '0x0000000000000000000000000000000000000000000000000000000000000001';

  // Create contract instance
  const myAssetContract = new ethers.Contract(
    '<myAssetAddress>',
    lsp8Artifact.abi,
    provider,
  );

  // Get the encoded asset metadata
  const tokenIdMetadata = await myAssetContract.getDataForTokenId(
    tokenID,
    ERC725YDataKeys.LSP4['LSP4Metadata'],
  );

  const erc725js = new ERC725(lsp4Schema);

  // Decode the metadata
  const decodedMetadata = erc725js.decodeData([
    {
      keyName: 'LSP4Metadata',
      value: tokenIdMetadata,
    },
  ]);
  console.log(decodedMetadata);
}
```

  </TabItem>

  <TabItem value="web3" label="web3">

```js
// ...

> if (isLSP8) {

Reviewer 2: I would suggest not wrapping the whole snipped in this `if (isLSP8)`. Its not really relevant to the guide imo as this guide only applies to LSP8

Reviewer 2: okay, maybe early return of something would clean it up a bit

    tokenID,
    ERC725YDataKeys.LSP4['LSP4Metadata'],
  );

  const erc725js = new ERC725(lsp4Schema);

  // Decode the metadata
  const decodedMetadata = erc725js.decodeData([
    {
      keyName: 'LSP4Metadata',
      value: tokenIdMetadata,
    },
  ]);

  console.log(decodedMetadata);
}
```

  </TabItem>

</Tabs>

<Details>
<summary>Show result</summary>

```js
{
    "key": "0x9afb95cacc9f95858ec44aa8c3b685511002e30ae54415823f406128b85b238e",
    "name": "LSP4Metadata",
    "value": {
      "verification": {
        "method": "keccak256(utf8)",
        "data": "0x9eefcdac3d60500619b075273c0371a6633d8f531179c882facd4f991281c658"
      },
      "url": "ipfs://QmeKNiTr4xfdHDUinGmYC4osu1ZHoFHDDj87WBSX5z4k7x"
    }
  }
```

</Details>

Based on the decoded metadata from the contract, you can continue retrieving the actual contents of the `Verifiable URI`. In this example, we will retrieve the file from an IPFS URL:

```js
// Prepare IPFS link to fetch
const contentID = decodedMetadata[0].value.url.replace('ipfs://', '');
const fileUrl = 'https://api.universalprofile.cloud/ipfs/' + contentID;

// Retrieve the metadata contents
const response = await fetch(fileUrl);

> const jsonMetadata = await response.text();

Reviewer 2: json probably makes more sense
```suggestion
const jsonMetadata = await response.json();
```
